subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmToolchainsPlugin.java

```diff
@@ -32,17 +37,28 @@
  * @since 7.6
  */
 @Incubating
-public class JvmToolchainsPlugin implements Plugin<Project> {
+public abstract class JvmToolchainsPlugin implements Plugin<Object> {
 
-    @Override
-    public void apply(Project target) {
-        target.getExtensions().create(JavaToolchainService.class, "javaToolchains", DefaultJavaToolchainService.class, javaToolchainQueryService);
-    }
-
-    private final JavaToolchainQueryService javaToolchainQueryService;
+    @Inject
+    protected abstract JavaToolchainQueryService getJavaToolchainQueryService();
 
     @Inject
-    public JvmToolchainsPlugin(JavaToolchainQueryService javaToolchainQueryService) {
-        this.javaToolchainQueryService = javaToolchainQueryService;
+    protected abstract DefaultJdksBlockForToolchainManagement getDefaultJdksBlockForToolchainManagement();
+
+    //TODO (#21082): update design docs about piggy-backing the old plugin
+
+    @Override
+    public void apply(Object target) {
```

Comment:
Of course we end up having to figure out where we are applied here. For some reason I though we could have two different methods.
Given this, I wonder if we should not have a `jvm-toolchains-settings` plugin or similar (based on the name of the existing one)
Or are others happy with the `instanceof` logic?

Reply:
I also think we shouldn't have done this merge. We don't really gain anything from it (except not having to come up with a new plugin name, but the one you are suggesting here seems perfectly fine to me) and we introduce a source of confusion: a plugin that does different things based on where it's applied. 

I'm going to extract a separate `jvm-toolchains-settings` plugin.

Reply:
I think I'm actually going with `jvm-toolchain-management` plugin.